### ISSUE TITLE
API Proposal: PollingWatcher

### ISSUE THREAD

**Issue by MisinformedDNA** [dotnet/corefxlab#2222]
```csharp
namespace System.IO.Enumeration  // was System.IO.FileSystem
{
    public class FileSystemMonitor : IDisposable  // was PollingWatcher
    {
        public FileSystemMonitor(string directory, EnumerationOptions options = null);
        public int PollingInterval { get; set; }
        // REMOVED: public PollingWatcher(string directory, bool includeSubdirectories = false, int pollingIntervalInMilliseconds = 1000);

        public FindPredicate ShouldRecursePredicate { get; set; }
        public FindPredicate ShouldIncludePredicate { get; set; }
        // REMOVED: public void AddExtension(string extension);

        public TraceSource Tracing { get; private set; }

        public void Start();
        public void Start(List<FileState> fileSystemState);
        public ReadOnlyCollection<FileState> GetSystemState();

        public Action Changed;
        public Action<FileChange[]> ChangedDetailed;
        public delegate bool FindPredicate(ref FileSystemEntry entry);

        public void Dispose();
    }

    public struct FileChange
    {
        public ReadOnlySpan<char> Path { get; }  // Full path
        // REMOVED: public string Name { get; }
        // REMOVED/REDUNDANT: public ReadOnlySpan<char> Directory { get; }

        public ChangeType ChangeType { get; }
    }

    public class FileState
    {
        public ReadOnlySpan<char> Path { get; }  // Full Path
        public DateTimeOffset LastWriteTimeUtc { get; }
        public long Length { get; }
        public override string ToString() { get; }

        public FileState(string path, DateTimeOffset lastWriteTimeUtc, long length);
    }
}
```

**Notes:**
 - Not a huge fan of the namespace `Enumeration` in general, but it seems like it belongs with `FileSystemEnumerable`.
 - Renamed to `FileSystemMonitor`. `PollingWatcher` is both not specific enough and is too similar to `FileSystemWatcher`, from which this is a pretty big departure.
 - Changed constructor to accept same parameters as `FileSystemEnumerable` (minus the transform).
 - Added similar predicate members from `FileSystemEnumerable` to pass through.
 - Added a `Start` overload to specify starting file set.
 - Add a method to retrieve file set.
    - The purpose of retrieving and setting the file set is so the user can decide to save state in case the system goes down, or only needs to be run on occassion. This helps prevent changes from being missed.
 - `FileState` is mutable as an internal version number changes; so it's a class

@stephentoub @JeremyKuhne @KrzysztofCwalina @ahsonkhan

**Comment by JeremyKuhne** [comment]
Thanks for pushing this prototype forward! Some of my initial observations:

> Not a huge fan of the namespace Enumeration in general, but it seems like it belongs with FileSystemEnumerable.

Common APIs stay in the root namespace (i.e. `System.IO`). `System.IO.Enumeration` was added for these advanced extensibility points, keeping the root namespace usability higher.

In that vein, you should consider a simpler, easier-to-use version of this that addresses the basic needs. Something that just takes `EnumerationOptions` and params of match strings would likely satisfy the vast majority of users. A more advanced, extensible version isn't precluded by having this.

> Renamed to FileSystemMonitor. PollingWatcher is both not specific enough and is too similar to FileSystemWatcher, from which this is a pretty big departure.

While it does have very different characteristics they both address the same need. From a discoverability perspective having Watcher in the name would be better. `PollingFileSystemWatcher` is pretty wordy, but not crazily so.

``` C#
    public struct FileChange
    {
        public ReadOnlySpan<char> Path { get; }  // Full path
        // REMOVED: public string Name { get; }
        // REMOVED/REDUNDANT: public ReadOnlySpan<char> Directory { get; }

        public ChangeType ChangeType { get; }
    }
```

The reason these were split into name and directory was for performance (notably allocation reduction). If you watch some directory structure you only have a small set of directory names and a lot of (usually much shorter) file names. Having the entire path as a string isn't always needed and can get crazy for deeply nested folders (Windows can go over 30000 characters deep!). Follow along with `FileSystemEntry` here for names. We _can_ make `ToFullPath()` a property instead (i.e. `FullPath`), but that should be cached internally.

Additionally, having spans in this isn't really useful unless the struct itself was a ref struct and was being passed as files are being enumerated.

> The purpose of retrieving and setting the file set is so the user can decide to save state in case the system goes down, or only needs to be run on occassion.

Why do this and not use `ISerializable`?

Would a `Stop()` cover the "run on occasion"? The current watcher has this capability by specifying a timeout and as such we should consider it.

> FileState is mutable as an internal version number changes; so it's a class

I'm not following this. Why wouldn't you keep this as a struct and simply reset the _version on import?

**Comment by MisinformedDNA** [comment]
>The reason these were split into name and directory was for performance (notably allocation reduction)

There may be other good reasons for keeping them separate, but wouldn't the allocation be the same? Directory and Path separately or just Path (FullPath). My thought on removing the directory is that we already know the root directory being monitored.

I don't have much experience with `ISerializable`, but I'm not sure where you would want me to use this.

The `_version` value is changed on every cycle, so it has to allow for changes. Import/export would ignore this value.

Let me think and experiment more on the rest.

**Comment by JeremyKuhne** [comment]
> There may be other good reasons for keeping them separate, but wouldn't the allocation be the same?

If you have 100 files in a directory that is 100 characters long that is nearly 20K of extra bytes. The key is that you keep 1 directory string for each directory. It is typical for watchers that VS uses to have tens of thousands of files being watched. Think in terms of 10Ks of files and 100s of directories for scaling as a minbar. I tested my enumeration changes against directories with over 500K files. :)

> I don't have much experience with ISerializable

You derive the class from it and implement the serializer methods and you can use any of the formatters to save/restore it. The doc page for ISerializable has good examples.

**Comment by MisinformedDNA** [comment]
I just assumed that all the directory strings would end up being duplicates of each other, but I guess if done correctly, that won't be the case. 👍

**Comment by KrzysztofCwalina** [comment]
Some feedback:

1. I don't like ShouldRecursePredicate and ShouldIncludePredicate. I think the public surface area should be just bool IncludeSubdirectories and then possibly have two virtual protected methods to do more elaborate filtering. Users who need it, would inherit from the base watcher.
2. I dont like public dependency on TraceSource. It's one of the many tracing technologies we have
3. I don't like that we expose the file system state. Do we really need it?
4. Why is the API using delegate fileds and not events? e.g. Changed
5. It's not clear what FileState.Length is
6. DateTime is better for UTC times (smaller). Do we need DTO?

**Comment by MisinformedDNA** [comment]
1. OK, that makes sense.
 2. Yea, that is pre-existing. Is any tracing necessary? I don't see any in `FileSystemWatcher`. What's the best practice here?
 3. My concern is that the machine reboots and then changes are missed. I'm not super fond of this, but I don't know how else to ensure that all changes are detected.
 4. Also pre-existing. You are referring to why we are using `Action` instead of `event`, right? I'm not sure why one is better than the other, but I can change that.
 5. I was reusing property names from `FileSystemEntry` and `FileInfo`. `Size` is more descriptive. `Length` is more consistent.
 6. @JeremyKuhne and you disagree here. DTO is consistent with `FileSystemEntry`, but I can't personally think of why we would need the Offset, besides consistency.
 7. @JeremyKuhne suggested `PollingFileSystemWatcher`. I'm good with that.

**Comment by MisinformedDNA** [comment]
> Common APIs stay in the root namespace (i.e. System.IO). System.IO.Enumeration was added for these advanced extensibility points, keeping the root namespace usability higher.

Considering the emergence of `Span<T>`, I was assuming that this was the replacement to the older way of doing things (i.e. not just advanced sceanrios).

**Comment by svick** [comment]
```c#
public void Start(List<FileState> fileSystemState);
public ReadOnlyCollection<FileState> GetSystemState();
```

This means that the output of `GetSystemState` and the input of `Start` are not compatible, i.e. I can't directly pass the output of `GetSystemState` as the input of `Start`. I think they should use compatible types.

**Comment by JeremyKuhne** [comment]
> DateTime is better for UTC times (smaller). Do we need DTO?

DateTimeOffset is safe, DateTime is not. @tarekgh can comment further if needed.

> Yea, that is pre-existing. Is any tracing necessary? I don't see any in FileSystemWatcher. What's the best practice here?

If diagnosing what is going on here would be useful we can have logging. You can follow along with what ArrayPool does. It might be nice to see all results (i.e. prefiltered) when logging is turned on.

> Considering the emergence of Span<T>, I was assuming that this was the replacement to the older way of doing things (i.e. not just advanced sceanrios).

That is useful feedback.

> This means that the output of GetSystemState and the input of Start are not compatible, i.e. I can't directly pass the output of GetSystemState as the input of Start. I think they should use compatible types.

I'd rather this use the existing ISerializable pattern than introducing a new one. :)

**Comment by KrzysztofCwalina** [comment]
I wonder if it would not be better for the state to be an opaque type, i.e. no members
Also, it might be good to make it "serializeble"

**Comment by MisinformedDNA** [comment]
`FileSystemWatcher` uses [WatcherChangeTypes](https://docs.microsoft.com/en-us/dotnet/api/system.io.watcherchangetypes?view=netcore-2.1):

```csharp
    [Flags]
    public enum WatcherChangeTypes
    {
        Created = 1,
        Deleted = 2,
        Changed = 4,
        Renamed = 8,
        All = 15
    }
```

But `PollingFileSystemWatcher` uses its own enum `ChangeType`:

```csharp
    public enum ChangeType : byte
    {
        Deleted = 0,
        Created = 1,
        Changed = 2,
    }
```

I'm not sure `PollingFileSystemWatcher` will ever use `Renamed` or `All`, but since `ChangeType` is a subset of `WatcherChangeTypes`,I was wondering if we should just get rid of `ChangeType`?

Note: The values per enum field are different, so this would be a breaking change, but, for itself, `PollingFileSystemWatcher` doesn't use the values.

**Comment by stephentoub** [comment]
My very high-level take on this is that, if we were to add a polling watcher, I'd want to see it added with as minimal amount of ceremony as possible into the existing FileSystemWatcher, rather than creating a whole brand new set of APIs, e.g.
```C#
public class FileSystemWatcher // existing class
{
    public void Start(bool polling) { ... } // or maybe arguments or an options bag relevant to polling
    ...
}
```
That would make it easy to switch back and forth between one based on the OS and one based on in-memory tracking and polling, while also keeping the surface area differences to a minimum.  It would also make it easier for platforms that don't support a native OS-based watching mechanism (or don't in all cases, such as when hitting inotify limits on Linux) to have a polling-based fallback already there and waiting.  I realize there are some new APIs being proposed for efficiency, e.g. some span-based members, but presumably we'd want those as well for the existing FileSystemWatcher, not just the new polling-based one.

My $.02 as a starting place.

**Comment by MisinformedDNA** [comment]
I initially proposed that `PollingFileSystemWatcher` should be in the `System.IO.Enumeration` namespace because that's where `FileSystemEnumerable` is. However, now I think it should go in the `System.IO` since that's where `FileSystemWatcher` is.

This way, someone in/using the `System.IO` namespace will see `FileSystemWatcher` and `PollingFileSystemWatcher` side by side.

**Comment by MisinformedDNA** [comment]
Based on @stephentoub's previous answer, I will put them in the same namespace

**Comment by JeremyKuhne** [comment]
@stephentoub The only mystery to me is what it might mean for a System.ComponentModel.Component to be a facade. It is just ignorance at this point- I really need to spend time looking at Component.

**Comment by stephentoub** [comment]
> The only mystery to me is what it might mean for a System.ComponentModel.Component to be a facade.

I'm not sure what you mean by that.  This can entirely be an implementation detail within FileSystemWatcher.

**Comment by JeremyKuhne** [comment]
> I'm not sure what you mean by that. This can entirely be an implementation detail within FileSystemWatcher.

Like I said, ignorance. Presumably there are no caveats to how you change a MarshalByRefObject in the BCL, but I just haven't spent much time with the concepts.

**Comment by MisinformedDNA** [comment]
**1.**

When I serialize, I want to know that I am not in the middle of a poll, or in other words, that the timer function is not in progress. Serializing while the timer function is in progress may lead to unexpected results.

One way to do this is to set the `PollingInterval` to `Timeout.Infinite`, which stops the polling for potential serialization. Once the serialization is finished, I can reset the PollingInterval to whatever and continue. However, if there isn't a change in the monitored directory, then the event will never trigger and then I can never be sure if the timer function is in progress or it's just waiting.

For this reason, I would like to add `public event EventHandler PollCompleted`, so I can know when a Timer is completed regardless of whether or not there was a change.

**2.**

Another scenario to consider is if someone changes the PollingInterval while the Polling is in progress. This could lead to two timer functions running side by side, instead of the desired single thread.

For these reasons, I think we need to add a setting `public bool IsPolling`.

**3.**

Another somewhat related concern is that `PollingInterval` accepts a `long` and relies on the user reading the comments to know that it is measured in milliseconds. Based on searches I've done, it seems like the more consistent approach would be to use a `TimeSpan`.

**Recommendations:**
 1. Add: `public event EventHandler PollCompleted`
 2. Add: `public bool IsPolling { get; set; }`
 3. Change: `public long PollingInterval` to return type `TimeSpan`

Thoughts?

**Comment by MisinformedDNA** [comment]
Proposed a simpler API to act as the base of `PollingFileSystemWatcher`. #2390
